### PR TITLE
Add routes for message translation #10

### PR DESCRIPTION
**Pull request by leiwang758**
### What this PR does (required):
- Handle the post message request, translate the message from the sender's primary language to all recipients' primary language, save them to a list and return as a response, and update the database.

### Screenshots / Videos (required):
Response from the route
image
Database Update
image

### CODE REVIEW

File: server/controllers/translation.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  var from_language = "";
+  var to_languages = [];
+
+  const conversation = await Conversation.findById(conversation_id);
+
+  // get sender's primary language
+  // and all other recipients' primary languages
+  for (let uid of conversation.users) {
+    let user = await User.findById(uid);
+    if (uid == user_id) {
```
Comment (dproc96) [inline server/controllers/translation.js:27]:
```suggestion
    if (uid === user_id) {
```
You may need to convert types to make this work but that's better practice

File: server/controllers/translation.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      from_language = user.primaryLanguage;
+    } else {
+      to_languages.push(user.primaryLanguage);
+    }
+  }
+
+  // translate
+  const translations = await Promise.all(
+    to_languages.map(async (l) => {
```
Comment (dproc96) [inline server/controllers/translation.js:36]:
```suggestion
    to_languages.map(async (toLanguage) => {
```
Best to be descriptive with variable names

File: server/routes/user.js [hunk 3]
```diff
@@ -7,11 +7,15 @@ const {
...
+const { postMessageTranslation } = require("../controllers/translation");
 
 router.route("/").get(protect, searchUsers);
 router.route("/conversations").get(protect, getUserConversations);
 router.route("/conversation/:userId").post(protect, postUserConversation);
 router.route("/messages/:conversationId").get(protect, getMessages);
 router.route("/message/:conversationId").post(protect, postMessage);
+router
+  .route("/translation/:conversationId")
+  .post(protect, postMessageTranslation);
```
Comment (dproc96) [inline server/routes/user.js:19]: What is the distinction between these two controllers? Shouldn't every message be translated when it gets posted?

**Review by dproc96 — COMMENTED** [review]
Had a couple of comments, additionally can you convert the snakecase variables to camelcase?

File: server/controllers/message.js [hunk 4]
```diff
@@ -26,38 +24,66 @@ exports.getMessages = asyncHandler(async (req, res) => {
...
-
+// Send a message and automatically translate it from sender's language to
+// all other recipients' languages and save the record
 exports.postMessage = asyncHandler(async (req, res) => {
-  const sender = req.user.id;
+  const userId = req.user.id;
   const conversationId = req.params.conversationId;
   const message = req.body.message;
+  var fromLanguage = "";
+  var toLanguages = [];
```
Comment (dproc96) [inline server/controllers/message.js:34]: Can you update these `var` to `const` and `let` accordingly?

File: server/controllers/message.js [hunk 5]
```diff
@@ -26,38 +24,66 @@ exports.getMessages = asyncHandler(async (req, res) => {
...
         },
       },
-      (err, doc) => {
+      (err) => {
         if (err) {
           return res.status(500).json({ error: err });
         } else {
           return res.status(200).json({
-            message,
+            text_to_translate: message,
```
Comment (dproc96) [inline server/controllers/message.js:83]: Should this be message instead of text_to_translate?

**Review by dproc96 — APPROVED** [review]
Looks good, just had two small comments. Also I noticed that the response object is using snakecase, can we switch that to camelcase?

File: server/controllers/message.js (hunk 4 above)
Comment (leiwang758) [reply server/controllers/message.js:34]: So the snakecase is due to the definition of the timestamp message model
```javascript
{ timestamps: { createdAt: "created_at", updatedAt: "updated_at" } }
```
I will change this to default which is camelcase

File: server/controllers/message.js (hunk 5 above)
Comment (leiwang758) [reply server/controllers/message.js]: Yes, I changed that as well, this is how the new response looks like
```javascript
{
            message,
            senderId: userId,
            chatId: conversationId,
            translations,
          }
```

**Pull request merged**